File: dsa_algo/patterns/sliding_window/8_replacing_ones.cpp
#include <iostream>
#include <bits/stdc++.h>

class replacing_ones {
    public:
        static int find_length(const std::vector<int>& arr, int k)
        {
            int l = 0;
            int max_length = 0;
            int max_ones_count = 0;
            for(int r = 0; r < arr.size(); r++)
            {

                if(arr[r] == 1)
                    max_ones_count++;
                if(r - l + 1 - max_ones_count > k)
                {
                    if(arr[l] == 1)
                        max_ones_count--;
                    l++;
                }
            max_length = std::max(max_length, r - l + 1);
            }
            return max_length;
        }
};

int main(int argc, char* argv[])
{
    std::cout<<replacing_ones::find_length(std::vector<int>{0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1}, 2)<<'\n';
    std::cout<<replacing_ones::find_length(std::vector<int>{0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1}, 3)<<'\n';
}

File: dsa_algo/patterns/sliding_window/8_replacing_ones.ts
const longest_substring = (arr: number[], k: number) =>
{
    let window_start = 0;
    let max_length = 0;
    let max_ones_count = 0;
    for(let window_end = 0; window_end < arr.length; window_end++)
    {
        if(arr[window_end] === 1)
            max_ones_count ++;
        if((window_end - window_start + 1 - max_ones_count) > k)
        {
            if(arr[window_start] === 1)
                max_ones_count -= 1;
            window_start += 1;
        }
        max_length = Math.max(max_length, window_end - window_start + 1);
    }
    return max_length;
}

console.log(longest_substring([0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1], 2));
console.log(longest_substring([0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1 ], 3));
